### PR TITLE
x86/64 U8S8 Gemm Precision Fix

### PR DESCRIPTION
**Pull request by chenfucn**
**Description**:
Add a graph optimization that convert u8s8 matrix multiplication to u8u8 if needed

**Motivation and Context**
In x86/64 platforms, specifically SSE4.1, AVX2 and AVX512 CPUs provide better performance computing u8s8 matrix multiplications. Unfortunately, the higher performance comes with value overflow problems, as described in:
https://www.intel.com/content/www/us/en/develop/documentation/onednn-developer-guide-and-reference/top/advanced-topics/nuances-of-int8-computations.html

In this change we added a session option "session.x64quantprecision" (default off). For operators that calls u8s8 matrix multiplications, e.g. QAttention, we convert them to u8u8 when the following conditions are all satisfied:

1. Current CPU is SSE4.1, AVX2 or AVX512 with no VNNI support
2. Session option "session.x64quantprecision" is on.
3. Constant weight tensor contains values outside of [-64, 63] range

Note that when weight tensor is not constant, QDQS8ToU8Transformer should already convert it to u8.

### CODE REVIEW

**Comment by yufenglee** [comment]
#11576

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 1]
```diff
@@ -84,6 +84,12 @@ static const char* const kOrtSessionOptionsConfigUseORTModelBytesDirectly = "ses
 // Available since version 1.11.
 static const char* const kOrtSessionOptionsQDQIsInt8Allowed = "session.qdqisint8allowed";
 
+// x64 AVX2/AVX512 has overflow problem with quantizied matrix multiplication with U8S8.
```
Comment (yufenglee) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:87]: >AVX512 [](http://example.com/codeflow?start=12&length=6)

nit: AVX512(without VNNI)

File: onnxruntime/core/mlas/lib/platform.cpp [hunk 2]
```diff
@@ -464,3 +465,17 @@
...
+#ifdef MLAS_TARGET_AMD64_IX86
+
+bool
+MLASCALL
+MlasPlatformU8S8Overflow(
+    void
+    )
+{
+    const auto& p = GetMlasPlatform();
+    return p.GemmU8U8Dispatch != p.GemmU8S8Dispatch;
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/platform.cpp:478]: They are different on SSE 4.1 too, but there is no overflow.

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+
+namespace onnxruntime {
+
+/**
+ * @brief Convert the source int8_t TensorProto to a uint8_t one if the tensor contains
+ *        value outside of [-64, 63]
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc:18]: >63 [](http://example.com/codeflow?start=33&length=2)

64 is also good. ORT quant tool uses range [-64, 64] for reduced range.

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+ * @param graph 
+ * @param op_node 
+ * @param weights_idx
+ * @param weight_zp_idx
+ * @return true when conversion happened.
+*/
+static bool ConvertS8WeightToU8(Graph& graph, Node& op_node, size_t weights_idx, size_t weight_zp_idx) {
+  auto& input_defs = op_node.MutableInputDefs();
+  if (input_defs.size() < std::max(weights_idx, weight_zp_idx) + 1) {
+    // TODO(fuchen): under what condition weight zp could be null? or not constant?
```
Comment (yufenglee) [inline onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc:73]: >zp could be null [](http://example.com/codeflow?start=49&length=16)

zp can be null if using default value(0).

File: onnxruntime/core/mlas/lib/platform.cpp (hunk 2 above)
Comment (chenfucn) [reply onnxruntime/core/mlas/lib/platform.cpp:478]: SSE4.1 does have overflow problem, it uses _mm_maddubs_epi16, which is [pmaddubsw](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_maddubs_epi16&ig_expand=4491)

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h (hunk 1 above)
Comment (chenfucn) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h]: fixed

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc (hunk 3 above)
Comment (chenfucn) [reply onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc]: fixed

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc (hunk 4 above)
Comment (chenfucn) [reply onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc]: fixed

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+struct OperatorWeightInfo {
+  const char* name;
+  std::vector<ONNX_NAMESPACE::OperatorSetVersion> versions;
+  const char* domain;
+  const size_t weights_idx;
+  const size_t weight_zp_idx;
+};
+
+static const std::vector<struct OperatorWeightInfo> s8_overflow_ops = {
```
[the remark this answers is not in the document]
Comment (chenfucn) [reply onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc]: done

File: onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+          // a single op_node can not possibly match more than 1 operator type
+          break;
+        }
+
+        if (graph_utils::IsSupportedOptypeVersionAndDomain(
+            op_node, "DynamicQuantizeLSTM", {1}, kMSDomain)) {
+          // This one has two set of quantized arguments
+          modified |= TryConvertDynamicQuantizeLSTM(op_node, graph);
+          break;
+        }
```
[the remark this answers is not in the document]
Comment (chenfucn) [reply onnxruntime/core/optimizer/avx2_weight_s8_to_u8.cc]: done

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**